src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -498,7 +498,7 @@ long getTokenStaleEpochSeconds() {
      *   <li>The agent need not be able to contact GitHub.
      * </ul>
      */
-    private Object writeReplace() {
+    protected Object writeReplace() {
```

Comment:
TBH this all seems a bit messy to me,
Why would this just not be in a Snapshot taker, and the refresh can make a call to the controller if on an agent or make a local call if it is on the controller...

Reply:
https://issues.jenkins.io/browse/JENKINS-73787

Reply:
> `DelegatingGitHubAppCredentials` does not extend `GitHubAppCredentials` and that's maybe why things started to get messy. It does not seem like you can snapshot a `GitHubAppCredentials` to a `DelegatingGitHubAppCredentials` with the current design ?

snapshot shout return (an implementation of) the interface type (because credential types should be written using interfaces as per the design doc) which in this case would be `StandardUsernamePasswordCredentials` so AFAICT without writing code this should be possible.
if code really does need a specific type for a GitHubApp credential (which is likely) then the GitHubAppCredential should not be a concrete type. ([doc](https://github.com/jenkinsci/credentials-plugin/blob/master/docs/implementation.adoc#implementing-a-new-credentials-type))

Reply:
(for the avoidance of doubt I am not blocking this PR, but I do not feel like the addition of debt onto the debt makes me want to approve it either)

Reply:
> if code really does need a specific type for a GitHubApp credential (which is likely)

No, it is just used as an implementation of `StandardUsernamePasswordCredentials` that happens to create a “password” on demand.

Reply:
TL;DR: Unless I am missing something obvious, this PR seems like the simplest way to fix the issue without causing more problems.

The motivation behind this change is to fix the behavior of a CloudBees-specific subclass of `GitHubAppCredentials` that comes from a plugin that implements a special `CredentialsProvider` that stores the relevant secrets externally (e.g. the private key). Right now, those credentials may not be used on agents directly.

The main problem is that we need multiple levels of dynamic behavior for GitHub app credentials:
* We need to support dynamic contextualization of `owner` via `Credentials.forRun` and `Connector.lookupScanCredentials` 
    * This requires us to use an instance of `GitHubAppCredentials` right up until the actual password is going to be used
    * For CloudBees, this means that any `CredentialsSnapshotTaker` for a `GitHubAppCredentials` _must_ translate to a type that is still a `GitHubAppCredentials` (which is not the case for `DelegatingGitHubAppCredentials`), otherwise `owner` inference totally breaks when credentials are sent from CloudBees CI Operations Center to controllers (this kind of problem was the motivation behind https://github.com/jenkinsci/github-branch-source-plugin/pull/796)
* We want to support dynamic refresh of the token from agents without the agents having access to the GitHub app secrets directly
    * This requires us to replace `GitHubAppCredentials` with a type like `DelegatingGitHubAppCredentials` when serializing the credentials to an agent

Making `writeReplace` `protected` here allows the CloudBees-specific type to use a `CredentialsSnapshotTaker` to produce an instance of `GitHubAppCredentials` where the external secrets have been retrieved, which does not break owner inference for the snapshot, and does not prevent `DelegatingGitHubAppCredentials` from being used for the credentials and the snapshot on agents.

If we try to use `CredentialsSnapshotTaker` here and delete `writeReplace`, things get more complicated. We need a way to tell in our `CredentialsSnapshotTaker` implementation whether the snapshot is happening because the credentials are being sent to an agent (the only case for OSS Jenkins) or is because the credentials are being sent from OC to controllers (only for CloudBees). This is not just a problem for the special `GitHubAppCredentials` subtype with CloudBees CI, it also applies to usage of plain `GitHubAppCredentials` in CloudBees CI (e.g. this kind of issue was the motivation behind https://github.com/jenkinsci/github-branch-source-plugin/pull/616). The problem though is that `Channel.current()` returns `null` from inside of `CredentialsSnapshotTaker.snapshot` (at least when checking `testAgentRefresh` here), so I don't really see any simple way to make this work without `writeReplace`.

Now, are there bigger changes that could be made to avoid this issue? Maybe. For example, if `DelegatingAppCredentials` was itself a subclass of `GitHubAppCredentials` and supported owner inference, then we could always translate directly to it in `CredentialsSnapshotTaker`. I am not sure if that can be done safely though given the need for agents to be able to refresh the token dynamically. Additionally, the CloudBees-specific subtype of `GitHubAppCredentials` could maybe be deleted with a redesign of the relevant plugin, but we'd still have CloudBees-specific issues with usage of regular `GitHubAppCredentials` if we tried to unify `GitHubAppCredentialsSnapshotTaker` and `GitHubAppCredentials.writeReplace`.